### PR TITLE
Multiagent fix

### PR DESCRIPTION
**Pull request by eugenevinitsky**
## What do these changes do?

These changes fix bugs in the PPO loss function for the multiagent case. Without these fixes, the multiagent code computes the loss incorrectly.

### CODE REVIEW

File: python/ray/rllib/models/action_dist.py [hunk 1]
```diff
@@ -8,108 +8,108 @@
...
-        ea1 = tf.exp(a1)
-        z0 = tf.reduce_sum(ea0, reduction_indices=[1], keep_dims=True)
-        z1 = tf.reduce_sum(ea1, reduction_indices=[1], keep_dims=True)
-        p0 = ea0 / z0
-        return tf.reduce_sum(p0 * (a0 - tf.log(z0) - a1 + tf.log(z1)),
-                             reduction_indices=[1])
-
-    def sample(self):
-        return tf.multinomial(self.inputs, 1)[0]
+  """Categorical distribution for discrete action spaces."""
```
Comment (richardliaw) [inline python/ray/rllib/models/action_dist.py]: hey @eugenevinitsky can we keep the 4 spaces for formatting? Thanks!

Comment (eugenevinitsky) [reply python/ray/rllib/models/action_dist.py]: Oh shoot, yes!

Comment (eugenevinitsky) [reply python/ray/rllib/models/action_dist.py]: Ignore above, I'm not sure why it's showing up without four spaces

Comment (eugenevinitsky) [reply python/ray/rllib/models/action_dist.py]: @richardliaw I think that should do it?

**Comment by ericl** [comment]
@eugenevinitsky any perf improvements on your benchmarks?

**Comment by eugenevinitsky** [comment]
If you mean the multiagent benchmarks, if you run the pendulum example with a batch size 4x larger than for your pendulum example, it should train in approximately the same time. The 4x comes from the fact that you're sampling from the variance of the Gaussian increasing. If you run it with GAE on, it trains in under 100 iterations. I'll submit some comparison plots with the pull request tomorrow.

File: python/ray/rllib/ppo/loss.py [hunk 2]
```diff
@@ -39,17 +40,49 @@ def __init__(
...
+                                           1 + config["clip_param"])
+                          for ratio_i in self.ratio]
+            self.surr = [tf.minimum(surr1_i, surr2_i) for surr1_i, surr2_i
+                         in zip(self.surr1, self.surr2)]
+            self.mean_policy_loss = tf.reduce_mean(-tf.add_n(self.surr))
+            self.surr = tf.add_n(self.surr)
+            self.entropy = tf.add_n(self.entropy)
+
+        else:
+            self.ratio = tf.exp(self.curr_dist.logp(actions) -
```
Comment (ericl) [inline python/ray/rllib/ppo/loss.py]: Is it possible to handle this as the single-list case, to avoid duplicating the code above?
e.g.
```
if not isinstance(curr_logp, list):
   curr_logp = [curr_logp]
   etc
```

File: python/ray/rllib/ppo/ppo.py [hunk 3]
```diff
@@ -92,8 +92,11 @@ class PPOAgent(Agent):
     _default_config = DEFAULT_CONFIG
 
     def _init(self):
+        self.num_agents = len(self.config["model"].
+                              get("custom_options", {})
+                              .get("multiagent_obs_shapes", [1]))
```
Comment (ericl) [inline python/ray/rllib/ppo/ppo.py]: this might not pass lint

File: python/ray/rllib/ppo/ppo.py [hunk 4]
```diff
@@ -181,19 +184,19 @@ def standardized(value):
...
-                    i, loss, policy_loss, vf_loss, kl, entropy))
+                    i, loss, policy_loss, vf_loss, np.sum(kl), entropy))
 
             values = []
             if i == config["num_sgd_iter"] - 1:
                 metric_prefix = "ppo/sgd/final_iter/"
                 values.append(tf.Summary.Value(
                     tag=metric_prefix + "kl_coeff",
-                    simple_value=self.kl_coeff))
+                    simple_value=self.kl_coeff[0]))
```
Comment (ericl) [inline python/ray/rllib/ppo/ppo.py]: mean?

File: python/ray/rllib/ppo/ppo.py [hunk 5]
```diff
@@ -181,19 +184,19 @@ def standardized(value):
...
             policy_loss = np.mean(policy_loss)
             vf_loss = np.mean(vf_loss)
-            kl = np.mean(kl)
+            kl = np.mean(kl, axis=0)
             entropy = np.mean(entropy)
             sgd_end = time.time()
             print(
                 "{:>15}{:15.5e}{:15.5e}{:15.5e}{:15.5e}{:15.5e}".format(
-                    i, loss, policy_loss, vf_loss, kl, entropy))
+                    i, loss, policy_loss, vf_loss, np.sum(kl), entropy))
```
Comment (ericl) [inline python/ray/rllib/ppo/ppo.py]: Should this be a mean?

File: python/ray/rllib/ppo/ppo_evaluator.py [hunk 6]
```diff
@@ -55,8 +55,10 @@ def __init__(self, registry, env_creator, config, logdir, is_remote):
 
         # Defines the training inputs:
         # The coefficient of the KL penalty.
+        num_agents = len(self.config["model"].get("custom_options", {}).
+                         get("multiagent_obs_shapes", [1]))
         self.kl_coeff = tf.placeholder(
-            name="newkl", shape=(), dtype=tf.float32)
+            name="newkl", shape=num_agents, dtype=tf.float32)
```
Comment (ericl) [inline python/ray/rllib/ppo/ppo_evaluator.py]: shouldn't this be `(num_agents,)`?

File: python/ray/rllib/models/action_dist.py [hunk 7]
```diff
@@ -135,23 +136,25 @@ def logp(self, x):
...
-        kl_list = np.asarray([distribution.kl(other_distribution) for
-                              distribution, other_distribution in
-                              zip(self.child_distributions,
-                                  other.child_distributions)])
-        return np.sum(kl_list)
+        kl_list = [distribution.kl(other_distribution) for
+                   distribution, other_distribution in
+                   zip(self.child_distributions,
+                       other.child_distributions)]
+        # return np.sum(kl_list)
```
Comment (ericl) [inline python/ray/rllib/models/action_dist.py]: can remove this

File: python/ray/rllib/models/action_dist.py [hunk 8]
```diff
@@ -135,23 +136,25 @@ def logp(self, x):
...
+        return kl_list
 
     def entropy(self):
         """The entropy of the action distribution."""
-        entropy_list = np.array([s.entropy() for s in
-                                 self.child_distributions])
-        return np.sum(entropy_list)
+        entropy_list = [s.entropy() for s in
+                        self.child_distributions]
+        # return np.sum(entropy_list)
```
Comment (ericl) [inline python/ray/rllib/models/action_dist.py]: also this

**Comment by ericl** [comment]
Looks good, some small comments. By the benchmarks I meant running a multiagent task (maybe one of the flow ones) before and after this change, to see if using the correctly summed PPO loss makes a difference.

File: python/ray/rllib/ppo/loss.py (hunk 2 above)
Comment (eugenevinitsky) [reply python/ray/rllib/ppo/loss.py]: Excellent point.

File: python/ray/rllib/ppo/ppo.py (hunk 4 above)
Comment (eugenevinitsky) [reply python/ray/rllib/ppo/ppo.py]: I'm actually unclear what we want to return here; the mean of the kl of a list of agents isn't informative.

**Review by ericl — APPROVED** [review]
Looks good. Could you run https://github.com/ray-project/ray/blob/master/python/ray/rllib/tuned_examples/regression_tests/cartpole-ppo.yaml and make sure it reaches 200 reward as before?

**Comment by eugenevinitsky** [comment]
Absolutely. Also, for reasons unknown to me the multiagent examples don't manage to learn anything unless I turn GAE on; even in the shared model case. I'd recommend not merging in until I have a satisfactory explanation for this?

**Comment by eugenevinitsky** [comment]
Confirmed:
== Status ==
Using FIFO scheduling algorithm.
Resources used: 0/8 CPUs, 0/0 GPUs
Result logdir: /Users/eugenevinitsky/ray_results/cartpole-ppo
TERMINATED trials:
 - PPO_CartPole-v0_0:	TERMINATED [pid=48042], 89 s, 30379 ts, 200 rew

**Comment by ericl** [comment]
I see, that's really odd, since IIUC this PR makes it so that the losses are entirely separate (unlike as before). Do you have some fast-running examples we can try out?

**Comment by ericl** [comment]
And pendulum is confirmed to work without GAE in non-multiagent mode right?

**Comment by eugenevinitsky** [comment]
This is a really good point, but non-multiagent pendulum does seem to work w/ GAE turned off.

**Comment by eugenevinitsky** [comment]
Further heads up, there seems to be another bug, in that the kl coefficients diverge after a bit for the shared models, indicating that either the gradients are not being propagated as expected in the shared models, or that the shared models are not correctly implemented.
